pkg/services/meta/search.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.uber.org/zap"
+)
+
+// Search selects up to count container's objects from the given container
+// matching the specified filters.
+func (m *Meta) Search(cID cid.ID, fs object.SearchFilters, fInt map[int]objectcore.ParsedIntFilter, attrs []string, cursor *objectcore.SearchCursor, count uint16) ([]client.SearchResultItem, []byte, error) {
+	m.stM.RLock()
+	s, ok := m.storages[cID]
+	m.stM.RUnlock()
+
+	if !ok {
+		m.l.Debug("skip search request due inactual container", zap.Stringer("cid", cID))
+		return nil, nil, nil
+	}
+
+	return s.search(fs, fInt, attrs, cursor, count)
+}
+
+func (s *containerStorage) search(fs object.SearchFilters, fInt map[int]objectcore.ParsedIntFilter, attrs []string, cursor *objectcore.SearchCursor, count uint16) ([]client.SearchResultItem, []byte, error) {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	if len(fs) == 0 {
+		return searchUnfiltered(s.db, cursor, count)
+	}
+
+	resHolder := objectcore.SearchResult{Objects: make([]client.SearchResultItem, 0, count)}
+	handleKV := objectcore.MetaDataKVHandler(&resHolder, &attrGetter{db: s.db}, nil, fs, fInt, attrs, cursor, count)
+
+	var rng storage.SeekRange
+	if cursor != nil {
+		rng.Prefix = cursor.PrimaryKeysPrefix
+		rng.Start = bytes.TrimPrefix(cursor.PrimarySeekKey, cursor.PrimaryKeysPrefix)
+	}
+	s.db.Seek(rng, func(k, v []byte) bool {
+		if cursor != nil && bytes.Equal(k, cursor.PrimarySeekKey) {
+			// points to the last response element, so go next
+			return true
+		}
+
+		return handleKV(k, v)
+	})
+
+	return resHolder.Objects, resHolder.UpdatedSearchCursor, resHolder.Err
+}
+
+// lock must be taken.
+func searchUnfiltered(st storage.Store, cursor *objectcore.SearchCursor, count uint16) ([]client.SearchResultItem, []byte, error) {
+	res := make([]client.SearchResultItem, 0, count)
+	var err error
+	var newCursor []byte
+
+	var rng storage.SeekRange
+	if cursor != nil {
+		rng.Prefix = cursor.PrimaryKeysPrefix
+		rng.Start = bytes.TrimPrefix(cursor.PrimarySeekKey, cursor.PrimaryKeysPrefix)
+	} else {
+		rng.Prefix = []byte{oidIndex}
+	}
+
+	st.Seek(rng, func(k, _ []byte) bool {
+		if bytes.Equal(k, cursor.PrimarySeekKey) {
+			return true
+		}
+		if len(res) == int(count) {
+			newCursor = res[len(res)-1].ID[:]
+			return false
+		}
+
+		if len(k) != oid.Size+1 {
+			err = fmt.Errorf("invalid meta bucket key (prefix 0x%X): unexpected object key len %d", k[0], len(k))
+			return false
+		}
+		res = append(res, client.SearchResultItem{ID: oid.ID(k[1:])})
+
+		return true
+	})
+
+	return res, newCursor, err
+}
+
+type attrGetter struct {
+	keyBuff []byte
+
+	db storage.Store
+}
+
+func (a *attrGetter) Get(oID []byte, attributeKey string) (attributeValue []byte, err error) {
+	defer func() {
+		a.keyBuff = a.keyBuff[:0]
```

Comment:
redundant since it is unknown whether the `keyBuf` will still be used

Reply:
i mean iiuc following code doesnt need this instruction to work properly. If so, it's no-op completely